
<style type="text/css">

<!--==========================================================================
	FTL macros 
	========================================================================== -->


<!--==========================================================================
	CSS styles 
	========================================================================== -->


<style type="text/css">

body {
	background-color: #dddddd;
}

/* Table styles. */

table {
  border-color: #000;
  border-spacing: 0px;
  border-style: solid;
  border-width: 1px;
  cell-spacing: 0px;
}

.noBorder {
  border-width: 0px;
}

td, th {
  font-family: Arial, Helvetica, sans-serif;
  font-size: 10pt;
  padding: 2px 0.5em;
  white-space: nowrap;
}

td.numeric {
  text-align: right;
}

th {
  background-color: #c0c0c0;
}

th.mainHeader {
  background-color: #808080;
  color: #ffffff;
  text-align: left;
}

th a {
  color: #000080;
  text-decoration: none;
}

th a:visited {
  color: #000080;
}

th a:active, th a:hover {
  color: #800000;
  text-decoration: underline;
}

.toc {
	border: 1px solid #aaa;
	background-color: #eeeeee;
	padding: 5px;
	font-size: 95%;
}

.note {
	border: 1px solid #aaa;
	background-color: #eeeeee;
	padding: 5px;
	font-size: 90%;
	text-align: left;
}

.main {
	margin-top: 15px;
	width: 1000px;
	text-align: left;
	background: #ffffff;
	border: 3px solid #777777;
}

.histo {
	table-layout:fixed;
	width:100%;
	border:1px solid #aaa;
	word-wrap:break-word;
}

</style>


<!--==========================================================================
	Index
	========================================================================== -->

<center>
<div class="main">

<center> <h3> SnpEff: Variant analysis </h3> </center>

<div style="margin-left: .5em">
<table class="toc"><tr><td>
	<center><b>Contents</b></center>
	<a href="#summary">Summary</a><br>
	<a href="#changeRateChr"> Change rate by chromosome</a><br>
	<a href="#changesByType">Variants by type</a><br>
	<a href="#effectsImpact"> Number of variants by impact </a><br>
	<a href="#effectsImpact"> Number of variants by functional class </a><br>
	<a href="#effects"> Number of variants by effect </a><br>
	<a href="#quality">Quality histogram</a><br>
	<a href="#coverage">Coverage histogram</a><br>
	<a href="#baseChages">Base change table</a><br>
	<a href="#tstv">Transition vs transversions (ts/tv)</a><br>
	<a href="#alleleFreq"> Frequency of alleles </a><br>
	<a href="#codonChanges"> Codon change table </a><br>
	<a href="#aaChanges"> Amino acid change table </a><br>
	<a href="#chrChanges"> Chromosome change plots </a><br>
	<a href="snpEff_genes.txt"> Details by gene </a><br>
</tr></td></table>
</div>


<!--==========================================================================
	Summary table 
	========================================================================== -->

<hr>
<a name="summary">
<center>
<b>Summary</b><p>

<table border=0>
	<tr bgcolor=ffffff> 
		<td valign=top> <b> Genome </b> </td>
		<td> BDGP5.66 </td>
	</tr>
	<tr bgcolor=dddddd> 
		<td valign=top> <b> Date </b> </td>
		<td> 2012-07-23 16:51 </td>
	</tr>
	<tr bgcolor=ffffff> 
		<td valign=top> <b> SnpEff version </b> </td>
		<td> <pre>SnpEff 3.1 (build 2012-07-20), by Pablo Cingolani</pre> </td>
	</tr>
	<tr bgcolor=dddddd> 
		<td valign=top> <b> Command line arguments </b> </td>
		<td> <pre>SnpEff  -v BDGP5.66 /Users/pablocingolani/snpEff/fly.bug.vcf 
</pre> </td>
	</tr>
	<tr bgcolor=ffffff> 
		<td valign=top> <b> Warnings </b> </td>
		   
		<td bgcolor="#ff0000"> 1 </td>
	</tr>	
	<tr bgcolor=dddddd>
		<td valign=top> <b> Number of lines (input file) </b> </td>
		<td> 1 </td>
	</tr>
	<tr bgcolor=ffffff>
		<td valign=top> <b> Number of variants (before filter) </b> </td>
		<td> 1 </td>
	</tr>
	<tr bgcolor=dddddd>
		<td valign=top> <b> Filter </b> </td>
		<td>  </td>
	</tr>
	<tr bgcolor=dddddd>
		<td valign=top> <b> Number of variants filtered out </b> </td>
		<td> 0 </td>
	</tr>
	<tr bgcolor=ffffff>
		<td valign=top> <b> Number of not variants <br>(i.e. reference equals alternative) </b> </td>
		<td> 0 </td>
	</tr>
	<tr bgcolor=dddddd>
		<td valign=top> <b> Number of variants processed <br> (i.e. after filter and non-variants) </b> </td>
		<td> 1 </td>
	</tr>
	<tr bgcolor=ffffff>
		<td valign=top> <b> Number of known variants <br>(i.e. non-empty ID) </b> </td>
		<td> 
		0
		( 0% ) 
		</td>
	</tr>
	<tr bgcolor=ffffff>
		<td valign=top> <b> Number of effects </b> </td>
		<td> 3 </td>
	</tr>
	<tr bgcolor=dddddd>
		<td valign=top> <b> Genome total length </b> </td>
		<td> 168,736,537 </td>
	</tr>
	<tr bgcolor=ffffff>
		<td valign=top> <b> Genome effective length </b> </td>
		<td> 23,011,544 </td>
	</tr>
	<tr bgcolor=dddddd>
		<td valign=top> <b> Change rate </b> </td>
		<td> 1 change every 23,011,544 bases </td>
	</tr>
</table>
<p>
</center>

<!--==========================================================================
	Change rate by chromosome
	========================================================================== -->

<hr> 
<a name="changeRateChr"> 

<center>
<b> Change rate details </b><p>

<table border=1>
	<tr><th> Chromosome </th><th> Length </th><th> Changes </th><th> Change rate </th></tr> 
		<tr>
			<td> 2L </td>
			<td class="numeric"> 23,011,544 </td>
			<td class="numeric"> 1 </td>
			<td class="numeric"> 23,011,544 </td>
		</tr> 
	<tr>
		<th> Total </th>
		<th class="numeric"> 23,011,544 </th>
		<th class="numeric"> 1 </th>
		<th class="numeric"> 23,011,544 </th>
	</tr> 
</table> 
</center>

<!--==========================================================================
	Changes by type
	========================================================================== -->

<hr> 
<a name="changesByType"> 
<center>
<b> Number changes by type</b><p>

<table border=1>
	<thead>
		<tr>
			<th> <b> Type   </b> </th>
			<th> <b> Total  </b> </th>
			<th> <b> Homo   </b> </th>
			<th> <b> Hetero </b> </th>
		</tr>
	</thead>
    <tbody>
	    <tr>
	    	<td> <b> SNP </b> </td>
	    	<td class="numeric" bgcolor="ffffff"> 1 </td>
	    	<td class="numeric" bgcolor="ffffff"> 1 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    </tr>
	    <tr>
	    	<td> <b> MNP </b> </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    </tr>
	    <tr>
	    	<td> <b> INS </b> </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    </tr>
	    <tr>
	    	<td> <b> DEL </b> </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    </tr>
	    <tr>
	    	<td> <b> MIXED </b> </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    </tr>
	    <tr>
	    	<td> <b> Interval </b> </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    </tr>
    </tbody>
    <tfoot>
	    <tr>
	    	<th><b>Total </b> </th>
	    	<th class="numeric"> 1 </th>
	    	<th class="numeric"> 1 </th>
	    	<th class="numeric"> 0 </th>
	    </tr>
    </tfoot>
</table>
</center>

<!--==========================================================================
	Effects by impact
	========================================================================== -->

<hr> 
<a name="effectsImpact"> 
<center>
<b> Number of effects by impact </b> <p> 

<table border=0>
	<thead>
		<tr> 
			<th><b> Type (alphabetical order)  </b></th>
			<th> &nbsp; </th>
			<th> Count </th>
			<th> Percent </th>
		</tr>
	</thead>
		<tr> 
			<td> <b> LOW </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#00ff00"> 1 </td> 
			<td class="numeric" bgcolor="#00ff00"> 33.333% </td>
		</tr>
		<tr> 
			<td> <b> MODIFIER </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#ff0000"> 2 </td> 
			<td class="numeric" bgcolor="#ff0000"> 66.667% </td>
		</tr>
</table><br>
<p>
</center>

<!--==========================================================================
	Effects by functional class
	========================================================================== -->

<hr> 
<a name="effectsFuncClass"> 
<center>
<b> Number of effects by functional class </b> <p> 

<table border=0>
	<thead>
		<tr> 
			<th><b> Type (alphabetical order)  </b></th>
			<th> &nbsp; </th>
			<th> Count </th>
			<th> Percent </th>
		</tr>
	</thead>
</table><br>
<p>

Missense / Silent ratio: </th><td class="numeric"> ?

</center>

<!--==========================================================================
	Effects stats
	========================================================================== -->

<hr> 
<a name="effects"> 
<center>
<b> Number of effects by type and region </b> <p> 

<table border=0>
	<tr>
		<th> Type </th>
		<th> Region </th>
	</tr>
	<tr>
		<td> <table border=0>
	<thead>
		<tr> 
			<th><b> Type (alphabetical order)  </b></th>
			<th> &nbsp; </th>
			<th> Count </th>
			<th> Percent </th>
		</tr>
	</thead>
		<tr> 
			<td> <b> SYNONYMOUS_CODING </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#00ff00"> 1 </td> 
			<td class="numeric" bgcolor="#00ff00"> 33.333% </td>
		</tr>
		<tr> 
			<td> <b> UPSTREAM </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#ff0000"> 2 </td> 
			<td class="numeric" bgcolor="#ff0000"> 66.667% </td>
		</tr>
</table><br>
 </td>
		<td> <table border=0>
	<thead>
		<tr> 
			<th><b> Type (alphabetical order)  </b></th>
			<th> &nbsp; </th>
			<th> Count </th>
			<th> Percent </th>
		</tr>
	</thead>
		<tr> 
			<td> <b> EXON </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#00ff00"> 1 </td> 
			<td class="numeric" bgcolor="#00ff00"> 33.333% </td>
		</tr>
		<tr> 
			<td> <b> UPSTREAM </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#ff0000"> 2 </td> 
			<td class="numeric" bgcolor="#ff0000"> 66.667% </td>
		</tr>
</table><br>
 </td>
	</tr>
</table>
<p>

<img src="http://chart.apis.google.com/chart?cht=bvs&chs=800x300&chbh=50,4,8&chxt=y,y,x,x&chtt=Variations&chxp=1,50.0|3,50.0&chxr=0,0.0,66.66666666666666|1,0.0,100.0|3,0.0,100.0&chco=00FFFF,FA8072,FF0000,A52A2A,FFA500,DEB887,E9967A,7FFFD4,808080&chdl=Upstream|5%27UTR|Exon|Splice+Donor|Intron|Splice+Acceptor|3%27UTR|Downstream|Intergenic&chxl=1:|%25|2:|Intergenic|Up|5%27UTR|Exon|Donor|Intron|Acceptor|Exon|3%27UTR|Down|Intergenic|3:|&chd=e:AA..AAAAAAAAAAAAAAAAAA,AAAAAAAAAAAAAAAAAAAAAA,AAAAAAgAAAAAAAgAAAAAAA,AAAAAAAAAAAAAAAAAAAAAA,AAAAAAAAAAAAAAAAAAAAAA,AAAAAAAAAAAAAAAAAAAAAA,AAAAAAAAAAAAAAAAAAAAAA,AAAAAAAAAAAAAAAAAAAAAA,AAAAAAAAAAAAAAAAAAAAAA" border=1><p>

</center>
<!--==========================================================================
	Quality plots
	========================================================================== -->

	<hr> 
	<a name="quality"> 
	<b> Quality: </b> 
	<p>
	<pre>
	<table class="histo">
		<tr> <th width=15%>Min</th><td>59</td> </tr>
		<tr> <th>Max</th><td>59</td> </tr>
		<tr> <th>Mean</th><td>59</td> </tr>
		<tr> <th>Median</th><td>29.5</td> </tr>
		<tr> <th>Standard deviation</th><td>0</td> </tr>
		<tr> <th>Values</th><td>59	</td> </tr>
		<tr> <th>Count</th><td>1	</td> </tr>
		</tr>
	</table>
		<img src="http://chart.apis.google.com/chart?cht=bvg&chd=e:..&chbh=8,4,2&chs=800x300&chxt=y,y,x,x&chxp=1,50.0|3,50.0&chtt=Quality+histogram&chxl=1:|Count|2:|59|3:|Quality&chxr=0,0.0,1.0|1,0.0,100.0|3,0.0,100.0"><br>
	</pre>

<!--==========================================================================
	Coverage
	========================================================================== -->

	<hr> 
	<a name="coverage"> 
	<b> Coverage:</b> 
	<p>
	<pre>
	<table class="histo">
		<tr> <th width=15%>Min</th><td>3</td> </tr>
		<tr> <th>Max</th><td>3</td> </tr>
		<tr> <th>Mean</th><td>3</td> </tr>
		<tr> <th>Median</th><td>1.5</td> </tr>
		<tr> <th>Standard deviation</th><td>0</td> </tr>
		<tr> <th>Values</th><td>3	</td> </tr>
		<tr> <th>Count</th><td>1	</td> </tr>
		</tr>
	</table>
		<img src="http://chart.apis.google.com/chart?cht=bvg&chd=e:..&chbh=8,4,2&chs=800x300&chxt=y,y,x,x&chxp=1,50.0|3,50.0&chtt=Coverage+histogram&chxl=1:|Count|2:|3|3:|Coverage&chxr=0,0.0,1.0|1,0.0,100.0|3,0.0,100.0"><br>
	</pre>

<!--==========================================================================
	InDels
	========================================================================== -->


<!--==========================================================================
	Base changes
	========================================================================== -->

<hr> 
<a name="baseChages"> 
<center>
<b> Base changes (SNPs) </b> <p>

<table border=1>
    <tr>
    	<td> &nbsp; </td>  <th> <b> A </b> </th> <th> <b> C </b> </th> <th> <b> G </b> </th> <th> <b> T </b> </th>
    </tr>
	    <tr> <th> <b> A </b> </th><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 1 </td> </tr>
	    <tr> <th> <b> C </b> </th><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td> </tr>
	    <tr> <th> <b> G </b> </th><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td> </tr>
	    <tr> <th> <b> T </b> </th><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td> </tr>
</table>
</center>

<p>

<!--==========================================================================
	Ts/Tv
	========================================================================== -->

<hr> 
<a name="tstv">
<center> <b> Ts/Tv (transitions / transversions) </b> <p> </center>

<small>
<b>Note:</b> Only SNPs are used for this statistic.<br>
<b>Note:</b> This Ts/Tv ratio is a 'raw' ratio. Some people prefer to use a ratio of rates, not observed events. In that case, you need to multiply by 2.0 (since there are twice as many possible transitions than transversions, E[Ts/Tv] ratio is twice the ratio of events).
</small>
<p>

<center>
<table border=1>
	<tr> <th> Transitions </th><td class="numeric"> 0 </td> </tr>
	<tr> <th> Transversions </th><td class="numeric"> 1 </td> </tr>
	<tr> <th> Ts/Tv ratio </th><td class="numeric"> 0 </td> </tr>
</table>
</center>
<p>

	<b>All variants:</b>
	<pre>Sample        :	Total
Transitions   :	0	0
Transversions :	2	2
Ts/Tv         :	0.000	0.000	
</pre>
	<p>
	<b>Only known variants</b> (i.e. the ones having a non-empty ID field):
	<pre>No results available (empty input?)</pre>
<p>

<!--==========================================================================
	Allele frequency
	========================================================================== -->

<hr> 
<a name="alleleFreq">
<center> <b> Frequency of alleles </b> <p> </center>

Note: Number of times an allele appears once (singleton), twice (doubletons), etc.<p>

	<b>All variants:</b>
	<img src="http://chart.apis.google.com/chart?cht=bvg&chd=e:..&chbh=8,4,2&chs=800x300&chxt=y,y,x,x&chxp=1,50.0|3,50.0&chtt=Alleles+frequencies&chxl=1:|Count|2:|2|3:|Alleles&chxr=0,0.0,1.0|1,0.0,100.0|3,0.0,100.0"><br>
	<table class="histo">
		<tr> <th width=15%>Min</th><td>2</td> </tr>
		<tr> <th>Max</th><td>2</td> </tr>
		<tr> <th>Mean</th><td>2</td> </tr>
		<tr> <th>Median</th><td>1</td> </tr>
		<tr> <th>Standard deviation</th><td>0</td> </tr>
		<tr> <th>Values</th><td>2	</td> </tr>
		<tr> <th>Count</th><td>1	</td> </tr>
		</tr>
	</table>
	<p>
	<b>Only known variants</b> (i.e. the ones having a non-empty ID field):
	<table class="histo">
		<tr> <th width=15%>Min</th><td>2,147,483,647</td> </tr>
		<tr> <th>Max</th><td>-2,147,483,648</td> </tr>
		<tr> <th>Mean</th><td>0</td> </tr>
		<tr> <th>Median</th><td>0</td> </tr>
		<tr> <th>Standard deviation</th><td>0</td> </tr>
		<tr> <th>Values</th><td></td> </tr>
		<tr> <th>Count</th><td></td> </tr>
		</tr>
	</table>
<p>


<!--==========================================================================
	Codon change table
	========================================================================== -->
	
<hr> 
<a name="codonChanges">
<center>
<b> Codon changes</b> <p>

	<div class="note">
		How to read this table: <br>
		- Rows are reference codons and columns are changed codons. E.g. Row 'AAA' column 'TAA' indicates how many 'AAA' codons have been replaced by 'TAA' codons.<br>
		- Red background colors indicate that more changes happened (heat-map).<br>
		- Diagonals are indicated using grey background color <br> 
		- WARNING: This table may include different translation codon tables (e.g. mamalian DNA and mitochondrial DNA).<br>
		<p>
	</div><p>

  
<table border=1>
	<thead>
		<tr> 
			<th> &nbsp; </th>
			 <th> TGG </th> 
		</tr>
	</thead>
			<tr> 
				<th> TGG </th>
					
						<th class="numeric"> 1 </th>
			</tr>
</table>
</center>

<!--==========================================================================
	Amino acid change table
	========================================================================== -->

<hr> 
<a name="aaChanges">
<center>
<b> Amino acid changes</b><p> 

	<div class="note">
		How to read this table: <br>
		- Rows are reference amino acids and columns are changed amino acids. E.g. Row 'A' column 'E' indicates how many 'A' amino acids have been replaced by 'E' amino acids.<br>
		- Red background colors indicate that more changes happened (heat-map).<br>
		- Diagonals are indicated using grey background color <br> 
		- WARNING: This table may include different translation codon tables (e.g. mamalian DNA and mitochondrial DNA).<br>
		<p>
	</div><p>

<table border=1>
	<thead>
		<tr> 
			<th> &nbsp; </th>
			 <th> W </th> 
		</tr>
	</thead>
			<tr> 
				<th> W </th>
					
						<th class="numeric"> 1 </th>
			</tr>
</table>
</center>

<!--==========================================================================
	Chromosome change table
	========================================================================== -->

<hr> 
<a name="chrChanges">
<center>
<b> Changes by chromosome</b><p> 

<center>
	<pre>
		<img src="http://chart.apis.google.com/chart?cht=lc&chs=800x300&chxt=y,y,x,x&chxp=1,50.0|3,50.0&chtt=Changes+histogram%3A+2L&chxr=0,0.0,0.0|1,0.0,100.0|3,0.0,100.0&chd=e:AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA&chxl=1:|Changes%2F100Kb|2:|100000||||||||||3000000||||||||||5900000||||||||||8800000||||||||||11700000||||||||||14600000||||||||||17500000||||||||||20400000|||||||||22950000|3:|Position"><br>
		Chromosome: 2L
	Position :	0	100000	200000	300000	400000	500000	600000	700000	800000	900000	1000000	1100000	1200000	1300000	1400000	1500000	1600000	1700000	1800000	1900000	2000000	2100000	2200000	2300000	2400000	2500000	2600000	2700000	2800000	2900000	3000000	3100000	3200000	3300000	3400000	3500000	3600000	3700000	3800000	3900000	4000000	4100000	4200000	4300000	4400000	4500000	4600000	4700000	4800000	4900000	5000000	5100000	5200000	5300000	5400000	5500000	5600000	5700000	5800000	5900000	6000000	6100000	6200000	6300000	6400000	6500000	6600000	6700000	6800000	6900000	7000000	7100000	7200000	7300000	7400000	7500000	7600000	7700000	7800000	7900000	8000000	8100000	8200000	8300000	8400000	8500000	8600000	8700000	8800000	8900000	9000000	9100000	9200000	9300000	9400000	9500000	9600000	9700000	9800000	9900000	10000000	10100000	10200000	10300000	10400000	10500000	10600000	10700000	10800000	10900000	11000000	11100000	11200000	11300000	11400000	11500000	11600000	11700000	11800000	11900000	12000000	12100000	12200000	12300000	12400000	12500000	12600000	12700000	12800000	12900000	13000000	13100000	13200000	13300000	13400000	13500000	13600000	13700000	13800000	13900000	14000000	14100000	14200000	14300000	14400000	14500000	14600000	14700000	14800000	14900000	15000000	15100000	15200000	15300000	15400000	15500000	15600000	15700000	15800000	15900000	16000000	16100000	16200000	16300000	16400000	16500000	16600000	16700000	16800000	16900000	17000000	17100000	17200000	17300000	17400000	17500000	17600000	17700000	17800000	17900000	18000000	18100000	18200000	18300000	18400000	18500000	18600000	18700000	18800000	18900000	19000000	19100000	19200000	19300000	19400000	19500000	19600000	19700000	19800000	19900000	20000000	20100000	20200000	20300000	20400000	20500000	20600000	20700000	20800000	20900000	21000000	21100000	21200000	21300000	21400000	21500000	21600000	21700000	21800000	21900000	22000000	22100000	22200000	22300000	22400000	22500000	22600000	22700000	22800000	22900000	23000000	
	Count    :	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	1	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	

	</pre>
</center>

<!--==========================================================================
	Changes by gene 
	========================================================================== -->

<center>
	<hr>
	<b> Details by gene </b><p>
	<p>
	<b><a href="snpEff_genes.txt">Here</a></b> you can find a tab-separated table.
</center>
</div>
</center>