<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>SnpEff</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="description" content="">
    <meta name="author" content="">

    <!-- Styles -->
    <link href="css/bootstrap.css" rel="stylesheet">
    <link href="css/bootstrap-responsive.css" rel="stylesheet">
    <link href="css/docs.css" rel="stylesheet">
    <link href="js/google-code-prettify/prettify.css" rel="stylesheet">
    <link href="css/my.css" rel="stylesheet">

    <!-- hTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="js/html5shiv.js"></script>
    <![endif]-->

    <!-- Fav and touch icons -->
    <link rel="apple-touch-icon-precomposed" sizes="144x144" href="ico/apple-touch-icon-144-precomposed.png">
    <link rel="apple-touch-icon-precomposed" sizes="114x114" href="ico/apple-touch-icon-114-precomposed.png">
      <link rel="apple-touch-icon-precomposed" sizes="72x72" href="ico/apple-touch-icon-72-precomposed.png">
                    <link rel="apple-touch-icon-precomposed" href="ico/apple-touch-icon-57-precomposed.png">
                                   <link rel="shortcut icon" href="ico/favicon.png">

  </head>

  <body data-spy="scroll" data-target=".bs-docs-sidebar">

    <!-- Navbar
    ================================================== -->
    <div class="navbar navbar-inverse navbar-fixed-top">
      <div class="navbar-inner">
        <div class="container-fluid">
          <button type="button" class="btn btn-navbar" data-toggle="collapse" data-target=".nav-collapse">
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
          </button>
          <a class="brand" href="#">SnpEff</a>
          <div class="nav-collapse collapse">
            <p class="navbar-text pull-right">
              <!--  Logged in as <a href="#" class="navbar-link">Username</a> -->
            </p>
            <ul class="nav">
              <li class="active"><a href="index.html">Home</a></li>
              <li><a href="download.html">Download</a></li>
              <li><a href="about.html">About</a></li>
            </ul>
          </div><!--/.nav-collapse -->
        </div>
      </div>
    </div>

<!-- Subhead
================================================== -->
<header class="jumbotron subhead" id="overview">
  <div class="container">
    <h1>SnpEff</h1>
    <p class="lead">Documentation</p>
  </div>
</header>


  <div class="container">

    <!-- Docs nav
    ================================================== -->
    <div class="row">
      <div class="span3 bs-docs-sidebar">
        <ul class="nav nav-list bs-docs-sidenav">
          <li><a href="#intro"><i class="icon-chevron-right"></i> Introduction</a></li>
          <li><a href="#install"><i class="icon-chevron-right"></i> Download & Install</a></li>
          <li><a href="#run"><i class="icon-chevron-right"></i> Running SnpEff </a></li>
          <li><a href="#seqData"><i class="icon-chevron-right"></i> Sequence data analysis </a></li>
          <li><a href="#input"><i class="icon-chevron-right"></i> Input files </a></li>
          <li><a href="#output"><i class="icon-chevron-right"></i> Output files </a></li>
          <li><a href="#eff"><i class="icon-chevron-right"></i> Effect prediction details </a></li>
          <li><a href="#"><i class="icon-chevron-right"></i> </a></li>
          <li><a href="#"><i class="icon-chevron-right"></i> </a></li>
          <li><a href="#"><i class="icon-chevron-right"></i> </a></li>
          <li><a href="#"><i class="icon-chevron-right"></i> </a></li>
        </ul>
      </div>
      <div class="span9">


        <!-- Introduction
        ================================================== -->
        <section id="intro">
          <div class="page-header">
            <h1>1. Introduction</h1>
          </div>
				<p class="lead">
				SnpEff is a variant annotation and effect prediction tool.
				It annotates and predicts the effects of genetic variants (such as amino acid changes).
				</p>

				<p> <b> Variants </b><br>
				By genetic variant we mean difference between a genome and a "reference" genome.
				As an example, imagine we are sequencing a "sample".
				Here "sample" can mean anything that you are interested in studing, from a cell culture, to a mouse or a cancer patient.
				<br>
				It is a standard procedure to compare your sample sequences against the corresponding "reference genome".
				For instance you may compare the cancer patient genome against the current "reference genome" (at the time of this writing, it is 'hg19').
				<p>
				In a typical sequencing experiment, you will find many places in the genome where your sample differs from the reference genome.
				These are called "genomic variants" or just "variants".
				<br>
				Typically, variants are catogorized as follows:
				<table class="table table-striped">
					<tr> <th> Name  </th> <th> What is means                    </th> <th> Example </th> </tr>
					<tr> <td> SNP   </td> <td> Single-Nucleotide Polymorphism   </td> <td> Reference = 'A', Sample = 'C' </td> </tr>
					<tr> <td> Ins   </td> <td> Insertion                        </td> <td> Reference = 'A', Sample = 'AGT' </td> </tr>
					<tr> <td> Del   </td> <td> Deletion                         </td> <td> Reference = 'AC', Sample = 'C' </td> </tr>
					<tr> <td> MNP   </td> <td> Multiple-nucleotide polymorphism </td> <td> Reference = 'ATA', Sample = 'GTC' </td> </tr>
				</table>
				This is not a comprehensive list, it is just to give you an idea.
				</p>

				<p> <b>Annotations</b><br>
				So, you have a huge file describing all the differences between your sample and the reference genome.
				But you want to know more about these variants than just their genetic coordinates. 
				E.g.: Are they in a gene? In an exon? Do they change protein coding? Do they cause premature stop codons?
				<br>
				SnpEff can help you answer all these questions.
				The process of adding this information about the variants is called "Annotation".
				<br>
				SnpEff provides several degrees of annotations, from simple (e.g. which gene is each variant affecting) to exremelly complex annotations (e.g. will this non-coding variant affect the expreesion of a gene?).
				It should be noted that the more complex the annotations, the more it relies in computational predictions.
				Such somputational predictions are sometimes incorrect. 
				<br>
				<i class="icon-warning-sign"></i> Results cannot be trusted blindly. SnpEff results may have to analyzed and independently validated.
				</p>

				<p> <b>Databases</b><br>
				In order to produce the annotations, SnpEff requires a database.
				We build these databases using informations from trusted resources.
				<br>
				Currently, there are pre-built database for over 2,500 reference genomes.
				This means that 99% of the cases are covered.
				<br>
				In some very rare ocasions, people need to build a database for an organism not currently supported (e.g. the genome is not publicly available).
				In most cases, this can be done and there is a section of this manual teaching how to build your own SnpEff database.
				<br>
				If you have problems adding you own organism, send me an email and I'll do my best to help you out.<p>

				<p> <b>Summary</b><br>
				A typical SnpEff use case would be:
				<ul>
					<li> Input:
					The inputs are predicted variants (SNPs, insertions, deletions and MNPs). 
					The input file is usually obtained as a result of a sequencing experiment, and it is usually in variant call format (VCF).
					<p>

					<li> Output:
					SnpEff analyzes the input variants. 
					It annotates the variants and calculates the effects they produce on known genes (e.g. amino acid changes).
					A list of effects and annotations that SnpEff can calculate can be found <a class="body" href="faq.html#What_effects_are_predicted?">here</a>.
				</ul>
				</p>

				<p> <b>Citing</b><br>
				If you are using SnpEff or SnpSift, please cite our work as shown <a href="SnpEff.html#citing">here</a>. Thank you!
				</p>
        </section>

        <!-- Download
        ================================================== -->
        <section id="install">
          <div class="page-header">
            <h1>2. Download &amp; Install</h1>
          </div>
				<p class="lead"> Download and installing SnpEff it pretty easy. 
				</p>
				
				<p> <b>System requirements </b><br>
				SnpEff requires that you have Java v1.6 or later installed (any modern operating system has it). <br>
				You'll probably need at least 2Gb of memory. 
				The amount of memory used can vary significantly depending on genome size and data analysis type you are doing.

				<p> <b>Downloading and Installing</b><br>
				In order to download and install the program, you can follow the directions <a href="download.html">here</a>.

				<p> <b> How is it licensed? </b><br>
				SnpEff is open source. It is released mainly as "LGPLv3". If you prefer another open source license, just contact me and I'll try to add your preferred license as well. <br>

				<p> <b>Source code </b><br>
				The project is hosted at SourceForge. <br>
				Here is the SVN command to check out the development version of the code:
<pre>
svn co https://snpeff.svn.sourceforge.net/svnroot/snpeff/SnpEffect/trunk
</pre>

				<p> <b>Configuration</b><br>
				The only configuration file is "snpEff.config".
				Most configuration parameters, are explained in the comments in the same config file, so I won't repeat the explanation here :-)
				<br>
				Usually you will only need to change the "data_dir" parameter.
				This parameter points to the "data" directory where you installed the tool (i.e. in unix systems, this is $HOME/snpEff/data).<p>
        </section>

        <!-- Running SnpEff
        ================================================== -->
        <section id="run">
          <div class="page-header">
            <h1>3. Running SnpEff</h1>
          </div>
				<p class="lead"> We show some basic examples how to use SnpEff.</p>

				<p><b>Command line vs Web interface</b><br>
				In order to run SnpEff you need to be comfortable running command from a command line terminal.
				If you are not, then it is probably a good idea to ask you systems adminitrator to install a <a class="body"class="body" href="http://usegalaxy.org">Galaxy</a> server and use the web interface.
				You can also use the open Galaxy server, but functionality may be limited and SnpEff versions may not be updated frequently.
				</p>

				<p><b>Usage example</b><br>
				In this example, we will annotate a variants file using the human reference genome.<br>

				As an input, we will use a <a href="http://www.1000genomes.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-41"> Variant Call Format (VCF) </a> file.
				VCF is the de-facto standard for genetic variants.
				<br>
				In this example we will use a file 'demo.1kg.vcf.gz', which is provided in the distribution.
<pre>
# Go to SnpEff install sirectory
cd ~/snpEff 

# Download human database 'hg19'
java -jar snpEff.jar download -v hg19

# Annotate 
java -Xmx2g -jar snpEff.jar hg19 -v demo.1kg.vcf > demo.1kg.snpeff.vcf
</pre>

				Here is an example of some entries in the annotated output file. 
				You can see the 'EFF' field was added, predicting STOP_GAINED protein changes:

<pre>
$ cat demo.1kg.snpeff.vcf | grep STOP_
1    889455    .    G    A    100.0    PASS    ...;EFF=STOP_GAINED(HIGH|NONSENSE|Cag/Tag|Q236*...
1    897062    .    C    T    100.0    PASS    ...;EFF=STOP_GAINED(HIGH|NONSENSE|Cag/Tag|Q141*...
1    900375    .    G    A    100.0    PASS    ...;EFF=STOP_GAINED(HIGH|NONSENSE|tGg/tAg|W578*...
</pre>
				Note: The real output was edited for readibility reasons.
				</p>

				<p><b>Runnig from another directory</b> <br>
				When you run SnpEff from a different directory than your install directory, you have to specify where the config file is located using the '-c' command line option.
<pre>
java -Xmx2g path/to/snpEff/snpEff.jar -c path/to/snpEff/snpEff.config GRCh37.70 path/to/snps.vcf
</pre>

				<p><b>Java memory options</b> <br>
				By default the amount of memory set by a java process is set too low.
				If you don't assigm more memory to the process, you will most likely have an "OutOfMemory" error.<br>

				You should set the amount of memory in your java virtual machine to, at least, 2 Gb. 
				This can be easily done using the Java command line option "-XmX".
				E.g. In this example I use 4Gb:
<pre>
# Run using 4 Gb of memory 
java -Xmx4G snpEff.jar hg19 path/to/your/files/snps.vcf
</pre>
				Note: There is no space between "-Xmx" and "4G".
				</p>

				<p><b>Running SnpEff in the Cloud </b> <br>
				You can run SnpEff in a "the Cloud" exactly the same way as running it on your local computer.
				You should not have any problems at all.
				<br>
				Here is an example of installing it and running it on an Amazon EC2 instance (virtual machine):<p>
<pre>
$ ssh -i ./aws_amazon/pcingola_aws.pem ec2-user@ec2-54-234-14-244.compute-1.amazonaws.com

       __|  __|_  )
       _|  (     /   Amazon Linux AMI
      ___|\___|___|


[ec2-user@ip-10-2-202-163 ~]$ wget http://sourceforge.net/projects/snpeff/files/snpEff_latest_core.zip
[ec2-user@ip-10-2-202-163 ~]$ unzip snpEff_latest_core.zip
[ec2-user@ip-10-2-202-163 ~]$ cd snpEff_3_1/
[ec2-user@ip-10-2-202-163 snpEff_3_1]$ java -jar snpEff.jar download -v hg19
00:00:00.000    Downloading database for 'hg19'
...
00:00:36.340    Done
[ec2-user@ip-10-2-202-163 snpEff_3_1]$ java -Xmx4G -jar snpEff.jar dump -v hg19 > /dev/null
00:00:00.000    Reading database for genome 'hg19' (this might take a while)
00:00:20.688    done
00:00:20.688    Building interval forest
00:00:33.110    Done.
</pre>

				<p>
				As you can see, it's very simple.

				<p><b>Command line options</b> <br>
				In order to see all available options, you can run 
				<pre>java -jar snpEff.jar -help</pre>
				We'll explain many of those options in detail in the following chapters.
				</p>

				<p><b>Loading the database</b> <br>
				One of the first thins SnpEff has to do is to load the database.
				Usually it takes frmo a few seconds to a couple of minutes, depending on database size.
				Complex databases, like human, require more time to load.
				After the database is loaded, SnpEff can analyze thousands of variants per second.
				</p>

        <!-- Sequence data analysis
        ================================================== -->
        <section id="seqData">
          <div class="page-header">
            <h1>4. Sequence data analysis</h1>
          </div>
				<p class="lead"> Here we show an example on how to get from Sequenceing data to an annotated variants file </p>

				<p><b>Sequencing data example</b> <br>
				<i class="icon-warning-sign"></i> This is an extremelly simplified version on how to analyze the data from scratch
				This is not meant to be a tutorial on sequencing analysis as it would be way beyond the scope of this handbook.
				<br>
				Let's assume you have sequence data in FASTQ format (file "s.fastq") and your reference genome is dm5.34 (fly genome)
<pre>
# Download the genome, uncompress and rename file
wget ftp://ftp.flybase.net/genomes/Drosophila_melanogaster/dmel_r5.34_FB2011_02/fasta/dmel-all-chromosome-r5.34.fasta.gz
gunzip dmel-all-chromosome-r5.34.fasta.gz
mv dmel-all-chromosome-r5.34.fasta dm5.34.fasta

# Create a genome index (we assume you installed BWA http://bio-bwa.sourceforge.net/)
bwa index -bwtsw dm5.34.fasta

# Map sequences to the genome: Create SAI file
bwa aln -bwtsw dm5.34.fasta s.fastq > s.sai

# Map sequences to the genome: Create SAM file
bwa samse dm5.34.fasta s.sai s.fastq > s.sam

# Create BAM file (we assume you installed SamTools http://samtools.sourceforge.net/)
samtools view -S -b s.sam > s.bam

# Sort BAM file (will create s_sort.bam)
samtools sort s.bam s_sort

# Create VCF file (BcfTools is part of samtools distribution)
samtools mpileup -uf dm5.34.fasta s_sort.bam | bcftools view -vcg - > s.vcf

# Analyze variants using snpEff
java -Xmx4g -jar snpEff.jar -vcf4 dm5.34 s.vcf > s_snpeff.txt
</pre>

				This highly simplified sequencing data analysis pipeline, has the common basic steps
				<ul>
					<li> Index the reference genome (bwa)
					<li> Map reads to reference genome (bwa)
					<li> Call varaints (bcftools)
					<li> Annotate variants (SnpEff)
				</ul>
				</p>
				
        <!-- Input
        ================================================== -->
        <section id="input">
          <div class="page-header">
            <h1>5. Input files </h1>
          </div>
				<p class="lead"> 
				Files used as input to SnpEff must comply with standard formats.
				Here we describe supported input data formats.
				</p>
				
				<p><b>VCF</b> <br>
				As we mentioned before, <a class="body"class="body" href="http://www.1000genomes.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-41">Variant Call Format (VCF)</a> is the recomended format for input files.
				This is the format used by the "1000 Genomes Project", and is currently considered the de-facto standard for genomic variants.
				Furthermore, most variant calling programs output VCF files, so it is also the simplest format to use.
				<br>
				In a nutshell, VCF format is tab-separated text file having the following columns:

				<ol>
					<li> Chromosome name
					<li> Position
					<li> Variant's ID
					<li> Reference genome
					<li> Alteranative (i.e. variant)
					<li> Quality score
					<li> Filter (whether or not the variant passed quality filters)
					<li> INFO : Generic information about this variant. SnpEff adds annotation information in this column.
				</ol>

				Here is an exampls of a few lines in a VCF file:
<pre>
#CHROM POS     ID        REF    ALT     QUAL FILTER INFO                    
20     14370   rs6054257 G      A       29   PASS   NS=3;DP=14;AF=0.5;DB;H2
20     17330   .         T      A       3    q10    NS=3;DP=11;AF=0.017   
</pre>
				Note that the first line is header information. Header lines start with '#'
				<p>

				<p><b>TXT:</b><br>
				<i class="icon-warning-sign"></i> TXT format is currently deprecated and discurraged.
				<br>
				TXT file format must be tab-separated, containing columns that correspond to:
				<ul>
					<li>chromosome_name
					<li>chromosome_position
					<li>reference sequence
					<li>changed sequence: A slash is used to separate two alleles
					<li>strand: {+,-} (optional)
					<li>quality (optional)
					<li>coverage (optional)
				</ul>

				This is an example of an input TXT file:

<pre>
5   140532    T    C       +
12  1017956   T    A       +    45    12
2   946507    G    C       +    23    8
14  19584687  C    T       -
19  66520     G    -G/-G   +
8   150029    A    */+T    +
</pre>

				<p><b>TXT: Representing an heterozygous variant</b><br> 
				<i class="icon-warning-sign"></i> TXT format is currently deprecated and discurraged.
				<br>
				When representing a heterozygous SNP you can use IUB codes: M=A/C, R=A/G, W=A/T, S=C/G, Y=C/T and K=G/T. 
				Indels are indicated by, for example, */+A, -A/* or +CC/-C. There is no difference between */+A or +A/*.

        <!-- Output
        ================================================== -->
        <section id="output">
          <div class="page-header">
            <h1>6. Output files </h1>
          </div>
				<p class="lead"> 
				Here we describe the format of SnpEff's output files and the information added.
				</p>

				<p><b>VCF output format</b><br> 
				As we mentioned in the previous chapter, VCF is the default input format.
				It is highly recomended to use VCF as input and output format, since it is a standard format that can be also used by other tools and software packages.
				Thus VCF makes it much easier to integrate genomic data processing pipelines.
				<br>
				SnpEff adds annotation information to the INFO field of a VCF file.
				The INFO field is the eigth column of a VCF file, see previous section for a quick example or take a look at the <a href="http://www.1000genomes.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-41">VCF specification</a> for details.
				<br>
				<br>
				Here is an example of a file before and after being annotated using SnpEff:
				<br>
				VCF file before annotations
<pre>
#CHROM POS     ID        REF    ALT     QUAL FILTER INFO                    
1	889455	.	G	A	100.0	PASS	AF=0.0005
1	897062	.	C	T	100.0	PASS	AF=0.0005
</pre>

				VCF file after being annotated using SnpEff
<pre>
#CHROM POS     ID        REF    ALT     QUAL FILTER INFO                    
1	889455	.	G	A	100.0	PASS	AF=0.0005;EFF=STOP_GAINED(HIGH|NONSENSE|Cag/Tag|Q236*|749|NOC2L||CODING|NM_015658|)
1	897062	.	C	T	100.0	PASS	AF=0.0005;EFF=STOP_GAINED(HIGH|NONSENSE|Cag/Tag|Q141*|642|KLHL17||CODING|NM_198317|)
</pre>
				A you can see, SnpEff added an 'EFF' tag to the INFO field (eigth column).

				<p>

				<p><b>VCF EFF field</b><br>
				<p>
				Effects information is added to the INFO field using an 'EFF' tag. 

				There can be multiple effects separated by comma. The format for each effect is:<p>
				Effect ( Effefct_Impact | Codon_Change | Amino_Acid_change | Gene_Name | Gene_BioType | Coding | Transcript | Exon [ | ERRORS | WARNINGS ] )<br>
				
				<center>
				<table class="table table-striped">
					<tr> <th>  EFF Sub-field </th> <th>  Meaning </th> </tr>
					<tr> <td>  Effect </td> <td>  Effect of this variant. See details <a class="body" href="#What_effects_are_predicted?">here</a>. </td> </tr>
					<tr> <td>  Effect impact </td> <td>  Effect impact {High, Moderate, Low, Modifier}. See details <a class="body" href="#How_is_impact_categorized?_(VCF_output)">here</a>. </td> </tr>
					<tr> <td>  Functional class </td> <td>  Functional class {NONE, SILENT, MISSENSE, NONSENSE}. </td> </tr>
					<tr> <td>  Codon_Change </td> <td>  Codon change: old_codon/new_codon  </td> </tr>
					<tr> <td>  Amino_Acid_change </td> <td>  Amino acid change: old_AA AA_position/new_AA (e.g. 'E30K') </td> </tr>
					<tr> <td>  Amino_Acid_length</td> <td>  Length of protein in amino acids (actually, transcription length divided by 3). </td> </tr>
					<tr> <td>  Gene_name </td> <td>  Gene name </td> </tr>
					<tr> <td>  Transcript_bioType </td> <td>  Transcript bioType, if available. </td> </tr>
					<tr> <td>  Gene_Coding </td> <td>  [CODING | NON_CODING]. This field is 'CODING' if any transcript of the gene is marked as protein coding. </td> </tr>
					<tr> <td>  Trancript </td> <td>  Transcript ID (usually ENSEMBL IDs) </td> </tr>
					<tr> <td>  Exon </td> <td>  Exon ID (usually ENSEMBL IDs) </td> </tr>
					<tr> <td>  Warnings </td> <td>  Any warnings or errors (not shown if empty). </td> </tr>
				</table>
				</center>
				<p>

				<p><b>VCF Header lines </b><br>
				SnpEff updates the header of the VCF file to reflect aditional fields.
				This is required by the VCF specification.
				SnpEff also adds the command line options used to annotate the file as well as SnpEff's version, so you can keep track of what exactly was done.

				<br>
				<br>
				Here is an example of some header lines added to an anotated file:
				<pre>
##SnpEffVersion="SnpEff 3.1m (build 2013-02-08)"
##SnpEffCmd="SnpEff  hg19 demo.1kg.vcf "
##INFO=&lt;ID=EFF,Number=.,Type=String,Description="Predicted effects for this variant.Format: 'Effect ( Effect_Impact | Functional_Class | Codon_Change | Amino_Acid_change| Amino_Acid_length | Gene_Name | Gene_BioType | Coding | Transcript | Exon [ | ERRORS | WARNINGS ] )' \"&gt;
</pre>

				<p><b>Mutiple effects per VCF line</b><br> 
				Usually there is more than one effect reported in each EFF field.
				<br>
				There are several reasons for this:
				<ul>
					<li> A variant can affect multple genes. E.g a variant can be DOWNSTREAM from one gene and UPSTREAM from another gene.
					<li> In complex organisms, genes usually have mutiple transcripts. So SnpEff reports the effect of a variant on each transcript.
					<li> A VCF line can have more then one variant. 
						 For instance if reference genome is 'G', but the sample has either 'A' or 'T', then this will be reported as one line, having multiple alternative variants. 
						 E.g.: 
<pre>
#CHROM  POS       ID   REF  ALT      QUAL  FILTER  INFO                    
1       889455    .    G    A,T      .     .       AF=0.0005
</pre>
						In this case SnpEff will report the effect of each variant on each gene and each transcript.
				</ul>
				</p>

				<p><b>Summary output file</b><br>
				SnpEff creates an additional output file showing overall statistics.
				This "stats" file is an HTML file which can be opned using a web browser.
				You can find an example of a 'stats' file <a href="1kg.html"> here</a>.
				<br>
				<br>
				The program performs some statistics and saves them to the file 'snpEff_summary.html' on the 
				directory where snpEff is being executed. You can see the file, by opeining it in your browser.
				<br>
				<br>
				<i class="icon-hand-right"></i>You can change the default location by using the '-stats' command line option. This also changes the location of the TXT summary file.
				<br>
				<br>
				E.g.: In the stats file, you can see coverage histogram plots like this one <img src="http://chart.apis.google.com/chart?cht=bvg&chbh=8,4,2&chs=800x300&chxt=y,y,x,x&chg=11.0,40.0,3,2&chxp=1,50.0|3,50.0&chtt=Coverage+histogram&chxr=0,1.0,409.0|1,0.0,100.0|3,0.0,100.0&chxl=1:|Count|2:|1||||||||||11||||||||||21||||||||||31||||||||||41||||||||||51||||||||||61||||||||||76||||||||||3:|Coverage&chd=e:DNBSEfGaI9MKRSWZfXnrlwmZuE1wvW630e6O..638y3r7g8JsKxRlHgAkeZmY9XCR7PXRSOuOFMzNcIUJmHrHrFIFIFxEfCkDNDNBSBSB7BSB7BSApApApBSBSApAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA">
				</p>

				<p> <b> Details about Hom/Het calculation (summary file and TXT output format) </b>
				<br>
				In older formats, there used to be one sample per file, so 
				identifying Hom/Het in a per variant basis made sense.
				<br>
				<br>
				In VCF files, where many samples are analyzed simultaneously, 
				identification of Hom/Het should be done in a per sample basis 
				instead of per variant basis. In an attempt to output that field I 
				used a majority vote (i.e. if most samples are Hom, I output 
				Hom, otherwise, Het). This is not strictly correct and, it 
				may also be confusing.
				<br>
				<br>
				For this reason, the output is empty when using multi-sample files, such as VCF. 
				Unless the file contains only one sample (i.e. only one genotype field)
				</p>


				<p><b>TXT output format</b><br> 
				<i class="icon-warning-sign"></i> TXT format is currently deprecated and discurraged.
				<br>
				The TXT output format consist of one line per effect. 
				This means that you usually get more than one line per SNP since the same SNP may affect several different transcripts in the same gene.
				<br>
				<p>
				The format consist of tab separated columns:
				<center>
				<table class="table table-striped">
					<tr> <th>  Column </th> <th>  Meaning </th> </tr>
					<tr> <td>  Chromosome </td> <td>  Chromosome name (usually without any leading 'chr' string) </td> </tr>
					<tr> <td>  Position </td> <td>  One based position </td> </tr>
					<tr> <td>  Reference </td> <td>  Reference </td> </tr>
					<tr> <td>  Change </td> <td>  Sequence change </td> </tr>
					<tr> <td>  Change type </td> <td>  Type of change {SNP, MNP, INS, DEL} </td> </tr>
					<tr> <td>  Homozygous </td> <td>  Is this homozygous or heterozygous {Hom, Het} </td> </tr>
					<tr> <td>  Quality </td> <td>  Quality score (from input file) </td> </tr>
					<tr> <td>  Coverage </td> <td>  Coverage (from input file) </td> </tr>
					<tr> <td>  Warnings </td> <td>  Any warnings or errors. </td> </tr>
					<tr> <td>  Gene_ID </td> <td>  Gene ID (usually ENSEMBL) </td> </tr>
					<tr> <td>  Gene_name </td> <td>  Gene name </td> </tr>
					<tr> <td>  Bio_type </td> <td>  BioType, as reported by ENSEMBL. </td> </tr>
					<tr> <td>  Trancript_ID </td> <td>  Transcript ID (usually ENSEMBL) </td> </tr>
					<tr> <td>  Exon_ID </td> <td>  Exon ID (usually ENSEMBL) </td> </tr>
					<tr> <td>  Exon_Rank </td> <td>  Exon number on a transcript </td> </tr>
					<tr> <td>  Effect </td> <td>  Effect of this variant. See details below. </td> </tr>
					<tr> <td>  old_AA/new_AA </td> <td>  Amino acid change </td> </tr>
					<tr> <td>  old_codon/new_codon </td> <td>  Codon change </td> </tr>
					<tr> <td>  Codon_Num(CDS) </td> <td>  Codon number in CDS </td> </tr>
					<tr> <td>  Codon_degenaracy </td> <td>  Codon degenaracy (see below). </td> </tr>
					<tr> <td>  CDS_size </td> <td>  CDS size in bases </td> </tr>
					<tr> <td>  Custom_interval_ID </td> <td>  If any custom interval was used, add the IDs here (may be more than one). </td> </tr>
				</table>
				</center>


				<p> 
				This is an example of an output file:<p>

				<small>
				<pre>
chr2L   35041   T       K       SNP     Het     50                      FBgn0051973     Cda5            FBtr0078164     FBgn0051973:11  11      SYNONYMOUS_CODING       Q/Q     CAA/CAA 332     5997    
chr2L   35041   T       K       SNP     Het     50                      FBgn0051973     Cda5            FBtr0078163     FBgn0051973:11  11      NON_SYNONYMOUS_CODING   Q/H     CAA/CAC 332     3120    
chr2L   200401  C       Y       SNP     Het     228                     FBgn0016977     spen            FBtr0078121     CG18497:10      10      STOP_GAINED     Y/*     TAC/TAG 5240    16602   
chr2L   194601  C       Y       SNP     Het     228                     FBgn0016977     spen            FBtr0078122     CG18497:6       6       NON_SYNONYMOUS_CODING   P/A     CCG/GCG 3473    16683   
chr2L   779563  A       R       SNP     Het     168                     FBgn0031277     CG13947         FBtr0078005     FBgn0031277:1   1       STOP_LOST       */Y     TAA/TAC 120     360   
chr2L   856021  *       +CGGAGGAGG/*    INS     Het     267                     FBgn0031288     CG13949         FBtr0078017     FBgn0031288:3   3       FRAME_SHIFT                     118     438     
chr2L   856890  C       Y       SNP     Het     201                     FBgn0029095     aru             FBtr0078031                     DOWNSTREAM: 43 bases   
chr2L   871556  G       R       SNP     Het     228                     FBgn0053526     PNUTS           FBtr0091487     CG33526:2       2       5PRIME_UTR: 18 bases from TSS  
chr2L   878267  T       W       SNP     Het     228             WARNING: Base is 'G' but SNP says 'T'.  FBgn0031292     CG15824         FBtr0078029     CG15824:14      14      3PRIME_UTR: 14 bases from transcript end  
				</pre>
				</small>

				<p> <i> Meaning of 'Degeneracy' column </i><br>
	  			Here is an explanation form Stephen Wright (Univ. Toronto), who requested to add this feature
	  
<pre>
	     "...a fourfold degenerate site would be a site where any change is synonymous. So the 
	     third codon position for the arginine CGN, is a fourfold degenerate site, as is the 
	     third codon position for valine, alanine, etc.
	     Similarly, all second positions of a codon are zerofold degenerate, since any change is 
	     nonsynonymous. Many first codon positions are also zerofold degenerate, however, for 
	     example, the first codon position of AGG is NOT zerofold, because a shift to CGG is a 
	     synonymous change."
</pre>
				</p>

				<br>
				<p> <b> Coordinates differences between VCT and TXT outputs </b>
				<br>
				<br>
				There are some situations when there seems to be a difference between TXT output coordinates and VCF output coordinates.
				This happens because sometimes VCF files have a coordinate that does not start exactly were the variant is. 
				Here is an example:
				<pre>
VCF line:                   "chr10   102770293       .       tgctgcggctgcggctgcggctacggctgcggct      tGCTGCGgctgcggctgcggctgcggctacggctgcggct"
INS reported by SnpEff:     "chr10   102770315       *       +GCGGCT INS"
</pre>
<p>

				At first glace, these two look different, since 102770293 is not the same as 102770315. 
				But if you look at the sequences:

<pre>
tgctgcggctgcggctgcggct      acggctgcggct    REF (from VCF file)
||||||||||||||||||||||      ||||||||||||
tgctgcggctgcggctgcggctgcggctacggctgcggct    ALT (from VCF file)
|      |         |    |||||| 
|      |         |    gcggct                INS (reported by SnpEff TXT format)
|      |         |    |
102770293        |    |                     Position (VCF)
       |         |    |
       102770300 |    |
                 |    |
                 102770310
                      |
                      102770315             Position (TXT)
</pre>
				So the two VCF and TXT representations are equivalent.
				</p>

				<br>
				<p> <b> Counting total number of effects of a given type </b>
				<br>
				Some people try to count the number of effects in a file by doing (assuming we want to count how many MODIFIER efefcts we have): 
				<br>
				<pre> grep -o MODIFIER output.eff.vcf | wc -l </pre>
				<br>
				This is incorrect because a VCF line can have multiple effects (e.g. when there are multiple transcripts in a gene).
				A proper way to count effects would be:

<pre>
cat output.eff.vcf \
	| cut -f 8 \
	| tr ";" "\n" \
	| grep ^EFF= \
	| cut -f 2 -d = \
	| tr "," "\n" \
	| grep MODIFIER \
	| wc -l
</pre>

				Brief explanation:

				<table class="table table-striped">
					<tr> <td> <tt>cut -f 8 </tt> </td> <td>  Extract INFO fields </td> </tr>
					<tr> <td> <tt>tr ";" "\n" </tt> </td> <td>  Expand each field into one line </td> </tr>
					<tr> <td> <tt>grep ^EFF= </tt> </td> <td>  Only keep 'EFF' fields </td> </tr>
					<tr> <td> <tt>cut -f 2 -d = </tt> </td> <td>  Keep only the effect data (drop the 'EFF=' part) </td> </tr>
					<tr> <td> <tt>tr "," "\n" </tt> </td> <td>  Expand effects to multiple lines </td> </tr>
					<tr> <td> <tt>grep MODIFIER | wc -l </tt> </td> <td>  Count the ones you want (in this example 'MODIFIER')</td> </tr>
				</table>
				</p>

        <!-- Effects
        ================================================== -->
        <section id="eff">
          <div class="page-header">
            <h1>7. Effect prediction details</h1>
          </div>
				<p class="lead">
				We describe the effects predicted by SnpEff
				</p>

				Here is a list of effects and some brief explanations:
				<center>
				<table class="table table-striped">
					<tr> <th> Effect </th><th> Note </th> <th> Example </th> </tr> 
					<tr> <td> INTERGENIC </td><td> The variant is in an intergenic region </td> </tr> 
					<tr> <td> UPSTREAM  </td><td> Upstream of a gene (default length: 5K bases) </td> </tr> 
					<tr> <td> UTR_5_PRIME  </td><td> Variant hits 5'UTR region </td> </tr> 
					<tr> <td> UTR_5_DELETED  </td><td> The variant deletes an exon which is in the 5'UTR of the transcript </td> </tr> 
					<tr> <td> START_GAINED  </td><td> A variant in 5'UTR region produces a three base sequence that can be a START codon. </td> </tr> 
					<tr> <td> SPLICE_SITE_ACCEPTOR  </td><td> The variant hits a splice acceptor site (defined as two bases before exon start, except for the first exon). </td> </tr> 
					<tr> <td> SPLICE_SITE_DONOR  </td><td> The variant hits a Splice donor site (defined as two bases after coding exon end, except for the last exon). </td> </tr> 
					<tr> <td> START_LOST   </td><td> Variant causes start codon to be mutated into a non-start codon. </td><td> aTg/aGg, M/R </td> </tr> 
					<tr> <td> SYNONYMOUS_START  </td><td> Variant causes start codon to be mutated into another start codon. </td><td> Ttg/Ctg, L/L (TTG and CTG can be START codons) </td> </tr> 
					<tr> <td> CDS  </td><td> The variant hits a CDS. </td> </tr> 
					<tr> <td> GENE  </td><td> The variant hits a gene. </td> </tr> 
					<tr> <td> TRANSCRIPT  </td><td> The variant hits a transcript. </td> </tr> 
					<tr> <td> EXON  </td><td> The vairant hits an exon. </td> </tr> 
					<tr> <td> EXON_DELETED  </td><td> A deletion removes the whole exon. </td> </tr> 
					<tr> <td> NON_SYNONYMOUS_CODING  </td><td> Variant causes a codon that produces a different amino acid </td><td> Tgg/Cgg, W/R </td> </tr> 
					<tr> <td> SYNONYMOUS_CODING  </td><td> Variant causes a codon that produces the same amino acid </td><td> Ttg/Ctg, L/L </td> </tr> 
					<tr> <td> FRAME_SHIFT  </td><td> Insertion or deletion causes a frame shift </td><td> An indel size is not multple of 3 </td> </tr> 
					<tr> <td> CODON_CHANGE  </td><td> One or many codons are changed </td><td> An MNP of size multiple of 3 </td> </tr> 
					<tr> <td> CODON_INSERTION  </td><td> One or many codons are inserted </td><td> An insert multiple of three in a codon boundary </td> </tr> 
					<tr> <td> CODON_CHANGE_PLUS_CODON_INSERTION  </td><td> One codon is changed and one or many codons are inserted </td><td> An insert of size multiple of three, not at codon boundary </td> </tr> 
					<tr> <td> CODON_DELETION  </td><td> One or many codons are deleted </td><td> A deletion multiple of three at codon boundary </td> </tr> 
					<tr> <td> CODON_CHANGE_PLUS_CODON_DELETION  </td><td> One codon is changed and one or more codons are deleted </td><td> A deletion of size multiple of three, not at codon boundary </td> </tr> 
					<tr> <td> STOP_GAINED  </td><td> Variant causes a STOP codon </td><td> Cag/Tag, Q/* </td> </tr> 
					<tr> <td> SYNONYMOUS_STOP  </td><td> Variant causes stop codon to be mutated into another stop codon. </td><td> taA/taG, */* </td> </tr> 
					<tr> <td> STOP_LOST  </td><td> Variant causes stop codon to be mutated into a non-stop codon </td><td> Tga/Cga, */R </td> </tr> 
					<tr> <td> INTRON  </td><td> Variant hits and intron. Technically, hits no exon in the transcript. </td> </tr> 
					<tr> <td> UTR_3_PRIME  </td><td> Variant hits 3'UTR region </td> </tr> 
					<tr> <td> UTR_3_DELETED  </td><td> The variant deletes an exon which is in the 3'UTR of the transcript </td> </tr> 
					<tr> <td> DOWNSTREAM  </td><td> Downstream of a gene (default length: 5K bases) </td> </tr> 
					<tr> <td> INTRON_CONSERVED  </td><td> The variant is in a highly conserved intronic region </td> </tr> 
					<tr> <td> INTERGENIC_CONSERVED </td><td> The variant is in a highly conserved intergenic region </td> </tr> 
					<tr> <td> INTRAGENIC </td><td> The variant hits a gene, but no transcripts within the gene</td> </tr> 
					<tr> <td> RARE_AMINO_ACID </td><td> The variant hits a rare amino acid thus is likely to produce protein loss of function </td> </tr> 
					<tr > <td> NON_SYNONYMOUS_START  </td><td> Variant causes start codon to be mutated into another start codon (the new codon produces a different AA). </td><td> Atg/Ctg, M/L (ATG and CTG can be START codons) </td> </tr> 
				</table> 
				</center>

				<p> <b>Details about Rare amino acid effect </b>
				<br>
				These are amino acids that occurs very rarely in an organism.
				For instance, humans are supposed to use 20 amino acids, but
				there is also one rare AA. Selenocysteine, single letter
				code 'U', appears roughly 100 times in the whole genome.
				The amino acid is so rare that usually it does not appear 
				in codon translation tables. It is encoded as UGA, which usually 
				means a STOP codon. Secondary RNA 
				structures are assumed to enable this special translation.
				<br>
				<br>
				A variant in one of these sites is likely to cause a loss of 
				function in the protein. E.g. in case of a Selenocysteine, a 
				loss of a selenium molecule is likely to cause loss of function.
				Put it simply, the assumption is that there is a great deal of trouble 
				to get that non-standard amino acid there, so it must be important.
				RARE_AMINO_ACID mark is used to show that special attention should 
				be paid in these cases.
				<br>
				<br>
				<i class="icon-warning-sign"></i> When the variant hits a RARE_AMINO_ACID mark, it is likely that 
				the 'old_AA/new_AA' field will be incorrect. This may happen because 
				the amino acid is not predictable using a codon table. 
				</p>

				<p><b>Effect impact</b><br> 
				Efefcts are categorized by 'impact': {High, Moderate, Low, Modifier}. 
				This are pre-defined categories to help users find more significant variants.
				<br>
				<i class="icon-warning-sign"></i> Categories must be used with care, they were created only to help and simplify the filtering process. 
				Obviously, there is no way to predict wether a "high impact" or a "low impact" variant is the one producing you phenotype of interest.
				<br>
				<br>
				Impact categories:
				<table class="table table-striped">
						<tr> <th> Impact </th><th> Effects </th> </tr> 
						<tr> <td> High </td>		
							<td> 
							SPLICE_SITE_ACCEPTOR<br>
							SPLICE_SITE_DONOR<br>
							START_LOST<br>
							EXON_DELETED<br>
							FRAME_SHIFT<br>
							STOP_GAINED<br>
							STOP_LOST<br>
							RARE_AMINO_ACID<br>
							</td> </tr> 
						<tr> <td> Moderate </td>	
							<td> 
							NON_SYNONYMOUS_CODING<br>
							CODON_CHANGE<br>
							CODON_INSERTION<br>
							CODON_CHANGE_PLUS_CODON_INSERTION<br>
							CODON_DELETION<br>
							CODON_CHANGE_PLUS_CODON_DELETION<br>
							UTR_5_DELETED<br>
							UTR_3_DELETED<br>
							</td> </tr> 
						<tr> <td> Low </td>		
							<td> 
							SYNONYMOUS_START<br>
							NON_SYNONYMOUS_START<br>
							START_GAINED<br>
							SYNONYMOUS_CODING<br>
							SYNONYMOUS_STOP<br>
							</td> </tr> 
						<tr> <td> Modifier </td>		
							<td> 
							UTR_5_PRIME<br>
							UTR_3_PRIME<br>
							REGULATION<br>
							UPSTREAM<br>
							DOWNSTREAM<br>
							GENE<br>
							TRANSCRIPT<br>
							EXON<br>
							INTRON_CONSERVED<br>
							INTRON<br>
							INTRAGENIC<br>
							INTERGENIC<br>
							INTERGENIC_CONSERVED<br>
							NONE<br>
							CHROMOSOME<br>
							CUSTOM<br>
							CDS<br>
							</td> </tr> 
				</table>
				</p>

				<p> <b> Loss of function and nonsense-mediated decay predictions </b>
				<br>
				<br>
				Loss of function ('LOF') and nonsense-mediated decay ('NMD') predictions can be activated by using '-lof' command line opion.
				Details on how these variants work, can be found in <a class="body" href="snpEff_lof_nmd.pdf">these slides </a>
				<br>
				Usage example:
<pre>
java –Xmx4g -jar snpEff.jar –v \
	-lof \
	GRCh37.68 \
	file.vcf.gz > file.eff.vcf
</pre>

				SnpEff adds 'LOF' and 'MND' tags to INFO fields (column 8 in VCF format).  LOF and NMD tags have the following format:
<pre>
Gene | ID | num_transcripts | percent_affected
</pre>
				Where:
				<table class="table table-striped">
					<tr> <th> Field            </th><th> Description </th> </tr>
					<tr> <td> Gene             </td><td> Gene name </td> </tr>
					<tr> <td> ID               </td><td> Gene ID (usually ENSEMBL) </td> </tr>
					<tr> <td> Num_transcripts  </td><td> Number of transcripts in this gene </td> </tr>
					<tr> <td> percent_affected </td><td> Percentage of transcripts affected by this variant. </td> </tr>
				</table>

				<br>
				Example: If we have this effect
<pre>
EFF=…, SPLICE_SITE_DONOR(HIGH||||639|ILDR2|protein_coding|CODING|ENST00000271417|1)
</pre>

				and the corresponding LOF tag is
<pre>
LOF=ILDR2|ENSG00000143195|7|1.00
</pre>

				The meaning of the LOF tag is:
				<table class="table table-striped">
					<tr> <th> Field            </th><th> Description </th> </tr>
					<tr> <td> Gene             </td><td> ILDR2 </td> </tr>
					<tr> <td> ID               </td><td> ENSG00000143195 </td> </tr>
					<tr> <td> Num_transcripts  </td><td> There are 7 transcripts in this gene </td> </tr>
					<tr> <td> percent_affected </td><td> 100% of transcripts are affected by this variant. </td> </tr>
				</table>
				</p>


        <!-- Filter output
        ================================================== -->
        <section id="eff">
          <div class="page-header">
            <h1>8. Output filters </h1>
          </div>
				<p class="lead">
				We describe how to filter the out or narrow down the number of results.
				</p>


				<p><b>Annotating only some intervals</b><br> 
				You can use the '-fi' option (a.k.a. '-filterInterval). For instance, let's assume you have an interval file 'intervals.bed':
<pre>2L	10000	10999
2L	12000	12999
2L	14000	14999
2L	16000	16999
2L	18000	18999 </pre>
				In order to get only changes that match your intervals, you can use the command:
<pre> java -Xmx4G -jar snpEff.jar -fi intervals.bed dm5.30 file.vcf</pre>
<b>Note:</b> You may specify multiple files by using '-fi' many times.

	     





				<p style="font-size: 14pt"> <b> <a class="body" name="gatk"> What is the adequate output for GATK? </a> </b></p>
				You should use the <b>'-o gatk'</b> command line option to avoid problems.
				The reason is that, even though GATK uses VCF, but sometimes new formats of the 'EFF' info field might cause trouble.
				This happens when switching from SnpEff v2.X to SnpEff v3.0, since the latest GATK version still uses an older format.<br>

				<hr> <center> <p style="font-size: 16pt"> <b> Output summary page </b></center></p>
				<p style="font-size: 14pt"> <b> <a class="body" name="Effects_by_type_vs_Effects_by_region"> "Effects by type" vs "Effects by region" </a> </b></p>
				SnpEff annotates variants. 
				Variants produce effect of difference "types" (e.g. NON_SYNONYMOUS_CODING, STOP_GAINED). 
				These variants affect regions of the genome (e.g. EXON, INTRON). <p>
				The two tables count how many effects for each type and for each region exists.
				<p>
				E.g.: In an EXON region, you can have all the following effect types:	NON_SYNONYMOUS_CODING, SYNONYMOUS_CODING, FRAME_SHIFT, STOP_GAINED, etc.
				<p>
				The complicated part is that some effect types affect a region that has the same name (yes, I know, this is confusing).
				<p>
				E.g.: In a UTR_5_PRIME region you can have UTR_5_PRIME and START_GAINED effect type. <br>
				This means that the number of both tables are not exactly the same, because the labels don't mean the same.
				See the next figure as an example
				<p>
				<img width=700 src="images/type_vs_region.png">
				<p>
				So the number of effects that affect a UTR_5_PRIME <b>region</b> is 206. Of those, 57 are effects <b>type</b> START_GAINED and 149 are effects <b>type</b> UTR_5_PRIME.
				<p>
				How exactly are effect <b>type</b> and effect <b>region</b> related? See the following table<p>
				<center>
				<table border=0>
					<tr bgcolor=aaaaaa> <th> Effect Type </th><th> ..affect Region </th> </tr> 
					 <tr bgcolor=eeeeee>
						<td>
						NONE<br>
						CHROMOSOME<br>
						CUSTOM<br>
						CDS<br>
						</td>
						<td>NONE</td>
					</tr>
					<tr bgcolor=dddddd>
						<td>
						INTERGENIC<br>
						INTERGENIC_CONSERVED<br>
						</td>
						<td>INTERGENIC</td>
					</tr>
					<tr bgcolor=eeeeee>
						<td>
						UPSTREAM<br>
						</td>
						<td>UPSTREAM</td>
					</tr>
					<tr bgcolor=dddddd>
						<td>
						UTR_5_PRIME<br>
						UTR_5_DELETED<br>
						START_GAINED<br>
						</td>
						<td>UTR_5_PRIME</td>
					</tr>
					<tr bgcolor=eeeeee>
						<td>
						SPLICE_SITE_ACCEPTOR<br>
						</td>
						<td>SPLICE_SITE_ACCEPTOR</td>
					</tr>
					<tr bgcolor=dddddd>
						<td>
						SPLICE_SITE_DONOR<br>
						</td>
						<td>SPLICE_SITE_DONOR</td>
					</tr>
					<tr bgcolor=eeeeee>
						<td>
						INTRAGENIC<br>
						START_LOST<br>
						SYNONYMOUS_START<br>
						NON_SYNONYMOUS_START<br>
						GENE<br>
						TRANSCRIPT<br>
						</td>
						<td>EXON or NONE</td>
					</tr>
					<tr bgcolor=dddddd>
						<td>
						EXON<br>
						EXON_DELETED<br>
						NON_SYNONYMOUS_CODING<br>
						SYNONYMOUS_CODING<br>
						FRAME_SHIFT<br>
						CODON_CHANGE<br>
						CODON_INSERTION<br>
						CODON_CHANGE_PLUS_CODON_INSERTION<br>
						CODON_DELETION<br>
						CODON_CHANGE_PLUS_CODON_DELETION<br>
						STOP_GAINED<br>
						SYNONYMOUS_STOP<br>
						STOP_LOST<br>
						RARE_AMINO_ACID<br>
						</td>
						<td>EXON</td>
					</tr>
					<tr bgcolor=eeeeee>
						<td>
						INTRON<br>
						INTRON_CONSERVED<br>
						</td>
						<td>INTRON</td>
					</tr>
					<tr bgcolor=dddddd>
						<td>
						UTR_3_PRIME<br>
						UTR_3_DELETED<br>
						</td>
						<td>UTR_3_PRIME</td>
					</tr>
					<tr bgcolor=eeeeee>
						<td>
						DOWNSTREAM<br>
						</td>
						<td>DOWNSTREAM</td>
					</tr>
					<tr bgcolor=dddddd>
						<td>
						REGULATION<br>
						</td>
						<td>REGULATION</td>
					</tr>
				</table>
				</center>


				<hr> <center> <p style="font-size: 16pt"> <b> Supporting genomes </b></center></p>

				<p style="font-size: 14pt"> <b> <a class="body" name="Where_is_the_human_genome?"> Where is the human genome?</a> </b></p>
				Human genome is now called 'GRCh37.XX' for ENSEMBL annotations, or 'hg19' for UCSC/RefSeq annotations. <br>
				Note that ENSEMBl provides better versioning (e.g GRCh37.63, GRCh37.64, GRCh37.65 ...) while UCSC has only one 'hg19' version which they might update. So your results using hg19 today might not be the same as using hg19 a year ago since UCSC might have updated the data, but the name is still hg19.

				<p style="font-size: 14pt"> <b> <a class="body" name="Which_genomes_are_supported?"> Which genomes are supported? </a> </b></p>
				There is a detailed list at the <a class="body" href="download.html">download</a> page.

				<p style="font-size: 14pt"> <b> <a class="body" name="How_do_I_add_support_for_XXX_genome?"> How do I build add support for XXX genome? How do I build a database?</a> </b></p>
				<a class="body"class="body" href="supportNewGenome.html">Here</a> is a detailed, step by step explanation.

				<p style="font-size: 14pt"> <b> <a class="body" name="How_do_I_add_support_for_regulatory_annotations?"> How do I add support for regulatory annotations? </a> </b></p>
				<a class="body"class="body" href="supportRegulation.html">Here</a> is a detailed, step by step explanation.

				<p style="font-size: 14pt"> <b> <a class="body" name="How_do_I_add_another_codon_table?"> How do I add another codon table?  </a> </b></p>
				You have to edit the <code>snpEff.config</code> file and add a parameter <code>"codon.Name_of_your_codon_table"</code> followed by a comma separaated list of <code>"CODON/AMINO_ACID"</code>. E.g.: 
<pre>
codon.Invertebrate_Mitochondrial:  TTT/F, TTC/F, TAC/Y, TAA/*, ATG/M+, ATG/M+, ACT/T, ...
</pre>
				Note that codons marked with '*' are STOP codons and codons marked with a '+' are START codons.<br>
				The next step is to tell snpEff to use that codon table for a given "chromosome" e.g. here we say the chromosome 'M' from fly genome (dm3) uses Invertebrate_Mitochondrial codon table:
<pre>
	dm3.M.codonTable : Invertebrate_Mitochondrial
</pre>

				...of course, chromosome 'M' is not a real chromosome, it is just a way to mark the sequence as mitochondrial DNA.

				<p style="font-size: 14pt"> <b> <a class="body" name="When_I_build_the_database_using_GFF_3_it_says_that_Exons_don't_have_s"> When I build the database using GFF 3 it says that Exons don't have sequences </a> </b></p>

				GFF3 files can have sequence information either in the same file or in a separate fasta file.<p>
				In order to add sequence information in the GFF file, you can do this:
<pre>
cat annotations.gff > genes.gff
echo "###"  >> genes.gff
echo "##FASTA"  >> genes.gff
cat sequence.fa  >> genes.gff
</pre>

				<p style="font-size: 14pt"> <b> <a class="body" name="I_get_zero_protein_coding_genes"> When building a database, I get zero protein coding genes</a> </b></p>
				When building a database, snpEff tries to find which transcripts are protein coding. This is done using the 'bioType' information.<br>
				The bioType information is not a standard GFF or GTF feature. So I follow ENSEMBL's convention of using the second column ('source') for bioType, as well as the gene_biotype attribute.<br>
 				If your file was not produced by ENSEMBL, it probably doesn't have this information. This means that snpEff doesn't know which genes are protein coding and which ones are not. <br>
				Having no information, snpEff will treat all genes as protein coding (assuming you have '-treatAllAsProteinCoding Auto' option in the command line, which is the default).<br>
 				So you will get effects as if all genes were protein coding, then you can filter out the irrelevant genes. unfortunately, this is the best I can do if there is no 'bioType' information <br>

				<p style="font-size: 14pt"> <b> <a class="body" name="I_get_too_many_warnings">When building a database, I get too many warnings</a> </b></p>
				There are plenty of GFF and GTF files that, unfortunately, do not follow the specification. 
				SnpEff usually complains about this, but tries hard to correct the problems. 
				So the database may be OK even after you see many warnings.<br>

				You can check the database to see if the features (genes, exons, UTRs) have been correctly incorporated, by taking a look at the database:

<pre>
java -jar snpEff.jar dump myGenome | less
</pre>

				<hr> <center> <p style="font-size: 16pt"> <b> Unexpected results </b></center></p>

				<p style="font-size: 14pt"> <b> <a class="body" name="Why_are_coordinates_changed_in_InDels?"> Why are coordinates changed in InDels?  </a> </b></p>

				This is not a bug, just a feature of VCF input files.
				Suppose that you have an InDel like this one

<pre>10      7797903 .       CAACTA  CAACTAACTA      214     .       INDEL;DP=94;AF1=1;AC1=2;DP4=0,0,90,0;MQ=53;FQ=-290      GT:PL:GQ        1/1:255,255,0:99</pre>

				After running snpEff, you get an output like this

<pre>10      7797909 *       +ACTA   INS     Hom     214     94              ENSG00000151657 KIN     protein_coding  ENST00000463666 exon_10_7797777_7798105 6       FRAME_SHIFT: ENST00000463666                            699</pre>

				Although it may seem like a change of coordinated (from 7797903 to 7797909) it not.
				If you look at the sequences, the first 6 bases are the same (i.e. the VCF file says that the INDEL is "CAACTA / CAACTAACTA" at position 7797903).
				So the actual insertion is "ACTA" a position 7797903 + 6 = 7797909.

				<p style="font-size: 14pt"> <b> <a class="body" name="When_I_look_at_the_UCSC_borwser_for_hg19,_it_doesn't_match_the_information_from_snpEff?"> When I look at the UCSC browser for hg19, it doesn't match the information from snpEff?  </a> </b></p>

				<b>WARNING:</b> Usage of hg19 genome is deprecated and discouraged, you should use GRChXX.YY instead<p>

				Reference sequence and annotations are made for an organism, version and sub-version. 
				For examples human genome, version 37, sub-version 63 would be called (GRCh37.63 or hg37.63 aka hg19.63).<br>
				UCSC doesn't specify sub-version. 
				They just say hg19.
				This annoying sub-version problem appeared often and, having reproducibility of results in mind, I dropped UCSC annotations in favor of ENSEMBL ones (they have clear versioning).<p>

				<p style="font-size: 14pt"> <b> <a class="body" name="When_I_run_SnpEff_from_GATK_I_only_get_one_effect!"> When I run SnpEff from GATK I only get one effect! </a> </b></p>
				Indeed,  the GATK team decided to only report the highest impact effect. This was done intentionally for the sake of brevity, in a 'less is more' spirit.<br>
				You can get the full effect by using snpEff independently (instead of using it from GATK).

				<p style="font-size: 14pt"> <b> <a class="body" name="Why_SnpEff_is_reporting_an_effect_that_doesn't_match_ENSEMBL_(web)?"> Why SnpEff is reporting an effect that doesn't match ENSEMBL (web)? </a> </b></p>
				Please remember that databases are updated often (e.g. by ENSEMBL), so if you are using an old database, you might get different effects.<br>
				<p>
				For example, this transcript ENST00000487462 changed from protein_coding in GRCh37.63 
<pre>
1       protein_coding  exon    1655388 1655458 .       -       .        gene_id "ENSG00000008128"; transcript_id "ENST00000487462"; exon_number "1"; gene_name "CDK11A"; transcript_name "CDK11A-013";
1       protein_coding  exon    1653905 1654270 .       -       .        gene_id "ENSG00000008128"; transcript_id "ENST00000487462"; exon_number "2"; gene_name "CDK11A"; transcript_name "CDK11A-013";
</pre>
...to processed_transcript in GRCh37.64:
<pre>
1       processed_transcript    exon    1655388 1655458 .       -       .        gene_id "ENSG00000008128"; transcript_id "ENST00000487462"; exon_number "1"; gene_name "CDK11A"; gene_biotype "protein_coding"; transcript_name "CDK11A-013";
1       processed_transcript    exon    1653905 1654270 .       -       .        gene_id "ENSG00000008128"; transcript_id "ENST00000487462"; exon_number "2"; gene_name "CDK11A"; gene_biotype "protein_coding"; transcript_name "CDK11A-013";
</pre>

				This means that you'll get different results for this transcript using sub-version 63 or 64. I assume that latest versions are improved, so I always encourage to upgrade.<p>
				As I mentioned before it might even be the case that latest released dataabse and the one shown on the web interface may be out of sync.<br>

				<p style="font-size: 14pt"> <b> <a class="body" name="Why_is_snpEff_reporting_a_SYNONYMOUS_and_a_NON_SYNONYMOUS_effect_on_te_same_gene?"> Why is snpEff reporting a SYNONYMOUS and a NON_SYNONYMOUS effect on the same gene? </a> </b></p>

				It is not uncommon for a gene to have more than one transcript (e.g. in human most genes have multiple transcripts). A variant (e.g. a SNP) might affect different transcripts in different ways, as a result of different reading frames. <p>
				For instance: 
<pre>
chr5 137622242 . C T . . EFF=NON_SYNONYMOUS_CODING(MODERATE|MISSENSE|Gaa/Aaa|E/K|CDC25C|protein_coding|CODING|ENST00000514017|exon_5_137622186_137622319),
                             SYNONYMOUS_CODING(LOW|SILENT|caG/caA|Q|CDC25C|protein_coding|CODING|ENST00000323760|exon_5_137622186_137622319),
                             SYNONYMOUS_CODING(LOW|SILENT|caG/caA|Q|CDC25C|protein_coding|CODING|ENST00000348983|exon_5_137622186_137622319),
                             SYNONYMOUS_CODING(LOW|SILENT|caG/caA|Q|CDC25C|protein_coding|CODING|ENST00000356505|exon_5_137622186_137622319),
                             SYNONYMOUS_CODING(LOW|SILENT|caG/caA|Q|CDC25C|protein_coding|CODING|ENST00000357274|exon_5_137622186_137622319),
                             SYNONYMOUS_CODING(LOW|SILENT|caG/caA|Q|CDC25C|protein_coding|CODING|ENST00000415130|exon_5_137622186_137622319),
                             SYNONYMOUS_CODING(LOW|SILENT|caG/caA|Q|CDC25C|protein_coding|CODING|ENST00000513970|exon_5_137622186_137622319),
                             SYNONYMOUS_CODING(LOW|SILENT|caG/caA|Q|CDC25C|protein_coding|CODING|ENST00000514555|exon_5_137622186_137622319),
                             SYNONYMOUS_CODING(LOW|SILENT|caG/caA|Q|CDC25C|protein_coding|CODING|ENST00000534892|exon_5_137622186_137622319)
</pre>

				in this example (it was divided into multiple lines for legibility), the first transcript ENST0000051401 has a NON_SYNONYMOUS effect, but all other transcripts have a SYNONYMOUS effect.
				<p>

				<hr> <center> <p style="font-size: 16pt"> <b> Other utilities </b></center></p>
				<p style="font-size: 14pt"> <b> <a class="body" name="countReads"> SnpEff countReads: Counting how many reads / bases from a BAM file hit a gene, transcript, exon, intron, etc. </a> </b></p>
				If you need to count how many reads (and bases) from a BAM file hit each genomic region, you can use 'countReads' utility.<br>
				The command line is quite simple. E.g. in order to count how many reads (from N BAM files) hit regions of the human genome, you simply run: <br>
<pre>
java -Xmx4g -jar snpEff.jar countReads GRCh37.68 readsFile_1.bam readsFile_2.bam ...  readsFile_N.bam > countReads.txt 
</pre>
				<p>
				The output is a TXT (tab-separated) file, that looks like this:
				<p>
<pre>
chr  start  end       type                IDs                         Reads:readsFile_1.bam  Bases:readsFile_1.bam  Reads:readsFile_2.bam  Bases:readsFile_2.bam ...
1    1      11873     Intergenic          DDX11L1                     130                    6631                   50                     2544
1    1      249250621 Chromosome          1                           2527754                251120400              2969569                328173439
1    6874   11873     Upstream            NR_046018;DDX11L1           130                    6631                   50                     2544
1    9362   14361     Downstream          NR_024540;WASH7P            243                    13702                  182                    9279
1    11874  12227     Exon                exon_1;NR_046018;DDX11L1    4                      116                    2                      102
1    11874  14408     Gene                DDX11L1                     114                    7121                   135                    6792
1    11874  14408     Transcript          NR_046018;DDX11L1           114                    7121                   135                    6792
1    12228  12229     SpliceSiteDonor     exon_1;NR_046018;DDX11L1    3                      6                      0                      0
1    12228  12612     Intron              intron_1;NR_046018;DDX11L1  13                     649                    0                      0
1    12611  12612     SpliceSiteAcceptor  exon_2;NR_046018;DDX11L1    0                      0                      0                      0
1    12613  12721     Exon                exon_2;NR_046018;DDX11L1    3                      24                     1                      51
1    12722  12723     SpliceSiteDonor     exon_2;NR_046018;DDX11L1    3                      6                      0                      0
1    12722  13220     Intron              intron_2;NR_046018;DDX11L1  22                     2110                   20                     987
1    13219  13220     SpliceSiteAcceptor  exon_3;NR_046018;DDX11L1    5                      10                     1                      2
1    13221  14408     Exon                exon_3;NR_046018;DDX11L1    82                     4222                   113                    5652
1    14362  14829     Exon                exon_11;NR_024540;WASH7P    37                     1830                   7                      357
1    14362  29370     Transcript          NR_024540;WASH7P            704                    37262                  524                    34377
1    14362  29370     Gene                WASH7P                      704                    37262                  524                    34377
1    14409  19408     Downstream          NR_046018;DDX11L1           122                    7633                   39                     4254
</pre>

				The columns are:
				<ul>
					<li> Column 1: Chromosome name
					<li> Column 2: Genomic region start
					<li> Column 3: Genomic region end
					<li> Column 4: Genomic region type (e.g. Exon, Gene, SpliceSiteDonor, etc.)
					<li> Column 5: ID (e.g. exon ID ; transcript ID; gene ID)
					<li> Column 6: Count of reads (in file readsFile_1.bam) intersecting genomic region.
					<li> Column 7: Count of bases (in file readsFile_1.bam) intersecting genomic region, i.e. each read is intersected and the resulting number of bases added.
					<li> Column ...:  (repeat count reads and bases for each BAM file provided)
				</ul>

				<hr> <center> <p style="font-size: 16pt"> <b> Troubleshooting </b></center></p>

				<p style="font-size: 14pt"> <b> <a class="body" name="I_get_an_OutOfMemoryError"> I get an OutOfMemoryError </a> </b></p>

				This means that your Java Virtual Machine (JVM) is running with low memory. You should increase the amount of memory assigned to the JVM by using the '-Xmx' command line option. For instance
<pre>
java -Xmx4G -jar snpEff.jar eff -v GRCh37.64 variants.vcf
</pre>

				Note that the amount of memory ("4G") is not separated by a space from the '-Xmx'. Be carefull not to exced the amount of physical memory on your computer, otherwise the process will start swaping to disc (you don't want that).

				<p style="font-size: 14pt"> <b> <a class="body" name="I_get_an_InvalidClassException"> I get an InvalidClassException </a> </b></p>

				This usually means that you are trying to use a database from another version . Binary databases are not backwards compatible (I hope to fix this soon).<br>
				The solution is to download the database which matches your SnpEff version. In order to get SnpEff's version, just run the program without any argument, and it will show the version in the first line, e.g.:
<pre>
$ java -Xmx64M -jar snpEff.jar

snpEff version 2.0.4 (build 2011-10-08), by Pablo Cingolani
Usage: snpEff [eff] [options] genome_version snp_file
   or: snpEff build [options] genome_version
   or: snpEff dump  [options] genome_version
   or: snpEff cds   [options] genome_version
</pre>

      </div>
    </div>

  </div>



    <!-- Footer
    ================================================== -->
      <footer>
        <p class="text-center">&copy; Pablo Cingolani 2013</p>
      </footer>


    <!-- Placed at the end of the document so the pages load faster -->
    <script type="text/javascript" src="http://platform.twitter.com/widgets.js"></script>
    <script src="js/jquery.js"></script>
    <script src="js/bootstrap-transition.js"></script>
    <script src="js/bootstrap-alert.js"></script>
    <script src="js/bootstrap-modal.js"></script>
    <script src="js/bootstrap-dropdown.js"></script>
    <script src="js/bootstrap-scrollspy.js"></script>
    <script src="js/bootstrap-tab.js"></script>
    <script src="js/bootstrap-tooltip.js"></script>
    <script src="js/bootstrap-popover.js"></script>
    <script src="js/bootstrap-button.js"></script>
    <script src="js/bootstrap-collapse.js"></script>
    <script src="js/bootstrap-carousel.js"></script>
    <script src="js/bootstrap-typeahead.js"></script>
    <script src="js/bootstrap-affix.js"></script>

    <script src="js/holder/holder.js"></script>
    <script src="js/google-code-prettify/prettify.js"></script>

    <script src="js/application.js"></script>



  </body>
</html>
